core-it-suite/src/test/java/org/apache/maven/it/MavenITmng6511OptionalProjectSelectionTest.java

```diff
@@ -0,0 +1,104 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.maven.it.util.ResourceExtractor;
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * This is a collection of test cases for <a href="https://issues.apache.org/jira/browse/MNG-6511">MNG-6511</a>,
+ * selecting and deselecting optional projects.
+ *
+ * @author Maarten Mulders
+ * @author Martin Kanters
+ */
+public class MavenITmng6511OptionalProjectSelectionTest extends AbstractMavenIntegrationTestCase
+{
+    private static final String RESOURCE_PATH = "/mng-6511-optional-project-selection";
+    private final File testDir;
+
+    public MavenITmng6511OptionalProjectSelectionTest() throws IOException
+    {
+        super( "[4.0.0-alpha-1,)" );
+        testDir = ResourceExtractor.simpleExtractResources( getClass(), RESOURCE_PATH );
+    }
+
+    public void testSelectExistingOptionalProfile() throws VerificationException
+    {
+        newVerifier( testDir.getAbsolutePath() ).executeGoal( "clean" );
+
+        Verifier verifier = newVerifier( testDir.getAbsolutePath() );
+        verifier.setLogFileName( "log-select-existing.txt" );
+        verifier.addCliOption( "-pl ?existing-module" );
+        verifier.executeGoal( "validate" );
+        verifier.verifyErrorFreeLog();
+        verifier.assertFilePresent( "existing-module/target/touch.txt" ); // existing-module should have been built.
+    }
+
+    public void testSelectExistingOptionalProfileByArtifactId() throws VerificationException
+    {
+        newVerifier( testDir.getAbsolutePath() ).executeGoal( "clean" );
+
+        Verifier verifier = newVerifier( testDir.getAbsolutePath() );
+        verifier.setLogFileName( "log-select-existing-artifact-id.txt" );
+        verifier.addCliOption( "-pl ?:existing-module" );
+        verifier.executeGoal( "validate" );
+        verifier.verifyErrorFreeLog();
+        verifier.assertFilePresent( "existing-module/target/touch.txt" ); // existing-module should have been built.
+    }
+
+    public void testSelectNonExistingOptionalProfile() throws VerificationException
+    {
+        newVerifier( testDir.getAbsolutePath() ).executeGoal( "clean" );
+
+        Verifier verifier = newVerifier( testDir.getAbsolutePath() );
+        verifier.setLogFileName( "log-select-non-existing.txt" );
+        verifier.addCliOption( "-pl ?non-existing-module" );
+        verifier.executeGoal( "validate" );
+        verifier.verifyErrorFreeLog();
+        verifier.assertFilePresent( "existing-module/target/touch.txt" ); // existing-module should have been built.
```

Comment:
How does this line relate to `non-exsisting-module`?

Reply:
Well, our idea was the following, if someone tries to select a module optionally and it does not exist, the build should just build everything. In other words `mvn validate -pl ?non-existing` has the same outcome as `mvn validate`.

Reply:
Alright, I have noticed that in other tests too. Maybe this special case should be documented where we have profile selection in maven-site. By defiinition, non-existing will cause the entire build. Take it or leave it.

Reply:
Yeah, sounds like a good idea! Updating the docs in maven-site is something we shouldn't forget, and especially this case.

Reply:
Please create a MNGSITE ticket for this.

Reply:
See [MNGSITE-444](https://issues.apache.org/jira/browse/MNGSITE-444).